nevergrad/optimization/recastlib.py

```diff
@@ -53,22 +54,167 @@ def _optimization_function(self, objective_function: tp.Callable[[tp.ArrayLike],
         # pylint:disable=unused-argument
         budget = np.inf if self.budget is None else self.budget
         best_res = np.inf
-        best_x: np.ndarray = self.current_bests["average"].x  # np.zeros(self.dimension)
+        best_x: np.ndarray = self.current_bests["average"].x
         if self.initial_guess is not None:
             best_x = np.array(self.initial_guess, copy=True)  # copy, just to make sure it is not modified
         remaining: float = budget - self._num_ask
         while remaining > 0:  # try to restart if budget is not elapsed
+            print(f"Iteration with remaining={remaining}")
             options: tp.Dict[str, tp.Any] = {} if self.budget is None else {"maxiter": remaining}
-            res = scipyoptimize.minimize(
-                objective_function,
-                best_x if not self.random_restart else self._rng.normal(0.0, 1.0, self.dimension),
-                method=self.method,
-                options=options,
-                tol=0,
-            )
-            if res.fun < best_res:
-                best_res = res.fun
-                best_x = res.x
+            if self.method == "SMAC2":
```

Comment:
please restructure all this, if there is little shared code, then maybe inheritance is a better path.
Also, this is way too big and should be split into smaller functions/building blocks to both ease understanding and testing.